### ISSUE TITLE
Simultaneous node and edge feature propagation through graph convolution layer

### ISSUE THREAD

**Issue by chi0tzp** [rusty1s/pytorch_geometric#1010]
## ❓ Questions & Help

I am looking for a graph convolution layer that would act on graphs (with given and constant topology -- i.e., a constant `edge_index`), and can take as input, along with node features, edge features (attributes), as well. However, I also need to output edge features as well. 

I'm relatively new to graph neural networks and I've not been aware of all the proposed methods, but I know that, for instance,  `nn.NNConv()` can additionally incorporate edge attributes during forward, but it doesn't output any. 

Also, I see that I might define an `nn.MetaLayer()`, where I could define a NodeModel and an EdgeModel. Would that be able to propagate node and edge features at the same time?

Eventually, after a number of such layers, I'd like to come up with a graph, where I'll have not only labels for its nodes, but also for its edges -- could I compute two losses there -- one for node classification and one for edge classification/regression? (Simillalry to what your example [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#learning-methods-on-graphs) shows for node classification).

Thank you very much.

**Comment by rusty1s** [comment]
Yeah, you can use of the `MetaLayer`, but it may just complicate things. The reason why this is mostly not integrated into PyG is that this is quite trivial to implement by yourself. If you want to modify edge features based on current edge features and node features, something like that will just do the trick:
```python
row, col = edge_index
new_edge_attr = self.mlp(torch.cat([edge_attr, x[col], x[row]], dim=-1))
```

**Comment by chi0tzp** [comment]
@rusty1s I see. So, I could just inherit from the base class and re-write NNConv (for instance) and add the edge attributes propagation as well (as above), if I understood correctly.

Thank you very much!

**Comment by rusty1s** [comment]
Alternatively, just write a basic `torch.nn.Module` that modifies your edge attributes in that way and call it between `NNConv`s.

**Comment by chi0tzp** [comment]
@rusty1s But that would treat edge attributes in a "fully-connected" way, right? I hoped that I could perform a graph-like convolution, just like with node attributes -- to aggregate edge features from the neighborhood of each node.

**Comment by rusty1s** [comment]
I see, than you need to work on two graphs, one that is the normal graph, and one that is called a [`LineGraph`](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.LineGraph) that transforms each edge into a node. You can then use all of the operators of PyG to work on this representation .

**Comment by chi0tzp** [comment]
@rusty1s Many thanks for taking the time to help -- really appreciate it. I'll try in the direction you propose shortly. I'll keep this issue open just in case I need some clarifications (I'll try not to bother you much). In any case, I feel that what I'm trying would be useful for others as well (maybe I'm wrong though).
